sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c

```diff
@@ -351,6 +351,31 @@ static int _fill_ipv6_hdr(gnrc_netif_t *netif, gnrc_pktsnip_t *ipv6)
             /* Otherwise leave unspecified */
         }
     }
+    else {
+        bool invalid_src;
+        int idx;
+
+        gnrc_netif_acquire(netif);
+        invalid_src = ((idx = gnrc_netif_ipv6_addr_idx(netif, &hdr->src)) == -1) ||
+            (gnrc_netif_ipv6_addr_get_state(netif, idx) != GNRC_NETIF_IPV6_ADDRS_FLAGS_STATE_VALID);
+        gnrc_netif_release(netif);
+        if (invalid_src) {
+            gnrc_pktsnip_t *icmpv6 = gnrc_pktsnip_search_type(ipv6,
```

Comment:
Does it make sense to surround all of this in `#if DEBUG` since it is only used for a debug message...  I don't know if the compiler would know about that.  Or just a separate static inline function?

Reply:
It is not just used for the debug message. It also triggers an error return.

Reply:
Oh, the return should be in the if below ^^... Mhhhh how did my tests work then yesterday oO

Reply:
I should clarify:
```c
if (invalid_src) {
#if DEBUG
    gnrc_pktsnip_t *icmpv6 = gnrc_pktsnip_search_type(ipv6, GNRC_NETTYPE_ICMPV6);
    icmpv6_hdr_t *icmpv6_hdr;
        if (icmpv6 != NULL) {
            icmpv6_hdr = icmpv6->data;
        }
        if ((icmpv6 == NULL) ||
            ((icmpv6_hdr->type != ICMPV6_RTR_SOL) &&
              (icmpv6_hdr->type != ICMPV6_NBR_SOL))) {
            DEBUG("ipv6: preset packet source address %s is invalid\n",
            ipv6_addr_to_str(addr_str, &hdr->src, sizeof(addr_str)));
    }
#endif
    return -EADDRNOTAVAIL;
}

Reply:
Yeah, the code should actually be.

```C
if (invalid_src) {
#if DEBUG
    gnrc_pktsnip_t *icmpv6 = gnrc_pktsnip_search_type(ipv6, GNRC_NETTYPE_ICMPV6);
    icmpv6_hdr_t *icmpv6_hdr;
    if (icmpv6 != NULL) {
        icmpv6_hdr = icmpv6->data;
    }
    if ((icmpv6 == NULL) ||
        ((icmpv6_hdr->type != ICMPV6_RTR_SOL) &&
          (icmpv6_hdr->type != ICMPV6_NBR_SOL))) {
        DEBUG("ipv6: preset packet source address %s is invalid\n",
            ipv6_addr_to_str(addr_str, &hdr->src, sizeof(addr_str)));
        return -EADDRNOTAVAIL;
    }
}
```

If it were right as is, I would agree with you. However, think about it: the check is not just done to produce the error message, but to exclude neighbor discovery messages used to determine the validity of an address.

Reply:
Fixed return's position

Reply:
:smiley: yup all good now!